dma: enable edma drivers for mcux

hakehuang opened the pull request:
edma rt series support. tested on mimxrt1060_evt/frdmk64f
MEMORY_NOCACHE is needed for rt
test slot need special set for kinetis

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

Depends on zephyrproject-rtos/hal_nxp#35

galak commented on dts/arm/nxp/nxp_rt.dtsi:
```diff
@@ -582,6 +582,37 @@
+			nxp,a_on;
+			reg = <0x400E8000 0x4000>,
+			        <0x400EC000 0x4000>;
+			interrupts = <0 0>,
```
Can format this like:

```
<0 0>, <1 0>, <2 0>, <3 0>
...
<15 0> <16>;
```

hakehuang replied:
ok

agansari commented on drivers/dma/dma_imx_rt.h:
```diff
@@ -0,0 +1,22 @@
+
+
+
+#endif /* DMA_STM32_H_*/
```
change /* DMA_STM32_H_*/

hakehuang replied:
ok

agansari commented on west.yml:
```diff
@@ -18,9 +18,12 @@ manifest:
     - name: upstream
       url-base: https://github.com/zephyrproject-rtos
+    - name:  nxp_dev
+      url-base: https://github.com/hakehuang/ 
```
change this before merge

hakehuang replied:
ok, I am discussion with maureen on how to create pull request for nxp_hal, once merged I will change align with mainline

MaureenHelm commented on drivers/clock_control/clock_control_mcux_ccm.c:
```diff
@@ -38,55 +36,59 @@ static int mcux_ccm_get_subsys_rate(struct device *dev,
+		break;
+#endif
+
+#ifdef CONFIG_DMA_IMX_RT
+	case IMX_CCM_EDMA_CLK:
+		*rate = CLOCK_GetIpgFreq();
 		break;
 #endif
```
Please make the clock control changes in two separate commits - 1 commit for the whitespace changes, 1 commit to add the EDMA clock.

hakehuang replied:
Ok

MaureenHelm replied:
These changes still appear in the same commit as the dma driver. Please move them to new commits; you can keep them in the same pull request.

hakehuang replied:
done

MaureenHelm commented on drivers/dma/CMakeLists.txt:
```diff
@@ -10,3 +10,4 @@ zephyr_library_sources_ifdef(CONFIG_DMA_DW		dma_dw.c)
 zephyr_library_sources_ifdef(CONFIG_DMA_NIOS2_MSGDMA	dma_nios2_msgdma.c)
 zephyr_library_sources_ifdef(CONFIG_DMA_SAM0		dma_sam0.c)
 zephyr_library_sources_ifdef(CONFIG_USERSPACE		dma_handlers.c)
+zephyr_library_sources_ifdef(CONFIG_DMA_IMX_RT		dma_imx_rt.c)
```
Can we call this `dma_mcux_edma.c` instead? Is the EDMA on imx rt similar enough to the EDMA on kinetis that we could use the same driver?

hakehuang replied:
let me check the clock settings on kinetis, and have a test on frdmk64f

hakehuang replied:
tested on frdmk64f, the driver can be shared with dts config, but the test application need some changes, I will provide a solution for review soon

MaureenHelm commented on drivers/dma/Kconfig.imx_rt:
```diff
@@ -0,0 +1,22 @@
+
+# Copyright (c) 2016 Intel Corporation
+# Copyright (c) 2019 Song Qiang <songqiang1304521@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
```
Add NXP copyright

hakehuang replied:
ok

MaureenHelm commented on drivers/dma/dma_imx_rt.c:
```diff
@@ -0,0 +1,498 @@
+};
+
+static __aligned(32) edma_tcd_t
+	tcdpool[DT_INST_0_NXP_IMX_EDMA_DMA_CHANNELS][CONFIG_DMA_TCD_QUEUE_SIZE];
```
why isn't this in `struct dma_imx_rt_data`?

hakehuang replied:
the TCD poll need special alignment 32 bytes address alignment, while the dma_data has no such requirement, just to keep the dma data compact. Any recommandation on how to use 32 bytes alignment in a structure? @MaureenHelm

MaureenHelm commented on drivers/dma/dma_imx_rt.c:
```diff
@@ -0,0 +1,498 @@
+#define DEV_EDMA_HANDLE(dev, ch)                                               \
+	((edma_handle_t *)(&(DEV_CHANNEL_DATA(dev, ch)->edma_handle)))
+
+static void _edma_callback(edma_handle_t *handle, void *param,
```
don't use leading underscores for driver functions

hakehuang replied:
ok

MaureenHelm commented on drivers/dma/dma_imx_rt.c:
```diff
@@ -0,0 +1,498 @@
+#endif
+}
+
+#if 0
+/* using fixed priority for RT is more user friendly */
+static void dam_imx_rt_set_channel_priority(struct device *dev, u32_t channel,
+					    struct dma_config *config)
+{
+	edma_channel_Preemption_config_t channel_priority;
+
+	channel_priority.enableChannelPreemption = true;
+	channel_priority.enablePreemptAbility = true;
+	channel_priority.channelPriority = config->channel_priority;
+	/*FIX ME channel need disabled first*/
+	EDMA_SetChannelPreemptionConfig(DEV_BASE(dev), channel,
+					&channel_priority);
+}
+#endif
```
remove dead code block

hakehuang replied:
ok

MaureenHelm commented on drivers/dma/dma_imx_rt.c:
```diff
@@ -0,0 +1,498 @@
+static void dma_imx_config_func_0(struct device *dev);
+
+static const struct dma_imx_rt_config dma_config_0 = {
+	.base = (DMA_Type *)DT_INST_0_NXP_IMX_EDMA_BASE_ADDRESS_0,
```
Please use the new DT_INST macros as in commit 4b9fe8a24e76b77e157c4d0d50f8c6aabadd9e87

hakehuang replied:
ok

MaureenHelm commented on drivers/dma/dma_imx_rt.c:
```diff
@@ -0,0 +1,498 @@
+	ARG_UNUSED(dev);
+
+/*install the dma error handle*/
+#if 1
```
remove

hakehuang replied:
ok

MaureenHelm commented on include/drivers/dma.h:
```diff
@@ -31,7 +31,8 @@ extern "C" {
 	MEMORY_TO_PERIPHERAL,
-	PERIPHERAL_TO_MEMORY
+	PERIPHERAL_TO_MEMORY,
+	PERIPHERAL_TO_PERIPHERAL /*only supported in NXP EDMA*/
```
please move all the changes in this file to a separate commit

hakehuang replied:
ok

MaureenHelm commented on tests/drivers/dma/chan_blen_transfer/src/test_dma.c:
```diff
@@ -26,8 +26,16 @@
 #define DMA_DEVICE_NAME CONFIG_DMA_0_NAME
 #define RX_BUFF_SIZE (48)
 
+#ifdef CONFIG_NOCACHE_MEMORY
```
please move all the changes in `tests/` to a separate commit

hakehuang replied:
ok

MaureenHelm commented on west.yml:
```diff
@@ -92,7 +92,8 @@ manifest:
     - name: hal_nxp
-      revision: 87850e41cd48e4472adbae03c9623ea349255d48
+      #revision: 87850e41cd48e4472adbae03c9623ea349255d48
+      revision: pull/35/head
```
adding `DNM` label until the hal pr gets merged

hakehuang replied:
ok

MaureenHelm commented on CODEOWNERS:
```diff
@@ -146,6 +146,8 @@
 /drivers/dma/*sam0*                       @Sizurka
 /drivers/dma/dma_stm32*                   @cybertale
+/drivers/dma/dma_mcux*                    @MaureenHelm
+/drivers/dma/Kconfig.imx_rt               @MaureenHelm
```
Once you have renamed the files to `Kconfig.mcux_edma` and `dma_mcux_edma.c`, you shouldn't need to add these lines anymore. They're covered by `/drivers/*/*mcux*`

hakehuang replied:
ok

MaureenHelm commented on boards/arm/mimxrt1060_evk/mimxrt1060_evk_defconfig:
```diff
@@ -14,3 +14,6 @@ CONFIG_CORTEX_M_SYSTICK=y
 CONFIG_ARM_MPU=y
+CONFIG_DMA=y
+CONFIG_DMA_IMX_RT=y
+CONFIG_NOCACHE_MEMORY=y
```
Instead of adding these lines to the board defconfig, update `soc/arm/nxp_imx/rt/Kconfig.defconfig.series` to enable the driver. Something like:
```
config DMA_MCUX_EDMA
        default y if HAS_MCUX_EDMA
        depends on DMA
```

hakehuang replied:
ok

hakehuang replied:
done

MaureenHelm commented on drivers/dma/CMakeLists.txt:
```diff
@@ -10,3 +10,5 @@ zephyr_library_sources_ifdef(CONFIG_DMA_DW		dma_dw.c)
 zephyr_library_sources_ifdef(CONFIG_DMA_NIOS2_MSGDMA	dma_nios2_msgdma.c)
 zephyr_library_sources_ifdef(CONFIG_DMA_SAM0		dma_sam0.c)
 zephyr_library_sources_ifdef(CONFIG_USERSPACE		dma_handlers.c)
+zephyr_library_sources_ifdef(CONFIG_DMA_IMX_RT		dma_mcux_edma.c)
```
remove this line

hakehuang replied:
ok

hakehuang replied:
done

MaureenHelm commented on drivers/dma/Kconfig:
```diff
@@ -48,4 +48,8 @@ source "drivers/dma/Kconfig.nios2_msgdma"
 
 source "drivers/dma/Kconfig.sam0"
 
+source "drivers/dma/Kconfig.imx_rt"
```
Remove this change

hakehuang replied:
ok

hakehuang replied:
done

MaureenHelm commented on drivers/dma/Kconfig.mcux:
```diff
@@ -0,0 +1,27 @@
+# Copyright (c) 2020, NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config DMA_MCUX
```
Please change this to `DMA_MCUX_EDMA` and rename the file to `Kconfig.mcux_edma`

hakehuang replied:
ok

MaureenHelm commented on drivers/dma/Kconfig.mcux:
```diff
@@ -0,0 +1,27 @@
+
+config DMA_MCUX
+	bool "Enable MCUX DMA driver"
+	depends on SOC_SERIES_KINETIS_K6X
```
Change this to `depends on HAS_MCUX_EDMA` and add an EDMA entry to `modules/Kconfig.mcux`. You'll then need to add `select HAS_MCUX_EDMA` to `soc/arm/nxp_imx/rt/Kconfig.soc` and `soc/arm/nxp_kinetis/k6x/Kconfig.soc`

hakehuang replied:
ok

hakehuang replied:
done

hakehuang committed:
dma: add EDMA test support for MCUX

for RT series CONFIG_NOCACHE_MEMORY need enabled
for Kinetis series CONFIG_DMA_TEST_SLOT_START for test
modify the target and dest alignment

frdmk64f need replace strlen to sizeof in irq callback
clean the unused receive buffer, so output is tidy

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
dma: add support for MCUX EDMA

1. expand PERIPHERAL_TO_PERIPHERAL
2. add slot size to 128 as i.MX-rt has 127 slot
3. add link_channel member in dma_conf which is used to
   support chain channel

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
clock_control: add EDMA clock support

add clock rate for RT series from ipg
Kinetis does not need additional clock added

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

MaureenHelm commented on drivers/clock_control/clock_control_mcux_ccm.c:
```diff
@@ -89,6 +89,12 @@ static int mcux_ccm_get_subsys_rate(struct device *dev,
+
+
+
+
```
why are you adding blank lines?

hakehuang replied:
> Please make the clock control changes in two separate commits - 1 commit for the whitespace changes, 1 commit to add the EDMA clock.

sorry this is not the way to add blank lines?

MaureenHelm replied:
No, I was referring to formatting changes, like adding spaces or tabs to fix alignment. It's easier to review when nonfunctional changes like this are in a separate commit.

hakehuang replied:
I see

MaureenHelm replied:
Thanks, it looks good

MaureenHelm reviewed (commented):
@galak please review the dts bindings

MaureenHelm commented:
@carlescufi I'd like to review the dma api changes in the api meeting next week

galak commented:
> @carlescufi I'd like to review the dma api changes in the api meeting next week

@erwango FYI since you guys are doing a bunch of DMA stuff on STM32 as well.

stephanosio commented:
> -:112: WARNING:SPDX_LICENSE_TAG: Missing or malformed SPDX-License-Identifier tag in line 1
> #112: FILE: drivers/dma/dma_mcux_edma.c:1:
> +/*

Please ignore these warnings (see #24921)

erwango commented on include/drivers/dma.h:
```diff
@@ -98,25 +99,26 @@ struct dma_block_config {
+ *     dma_slot             [ 0 : 6 ]   - which peripheral and direction
  *                                        (HW specific)
- *     channel_direction    [ 6 : 8 ]   - 000-memory to memory, 001-memory to
+ *     channel_direction    [ 7 : 9 ]   - 000-memory to memory, 001-memory to
```
Nit: let's use 1 value per line in field comment

hakehuang replied:
ok

erwango commented on drivers/dma/Kconfig.mcux_edma:
```diff
@@ -0,0 +1,29 @@
+	help
+	  number of TCD in a queue for SG mode
+
+config DMA_TEST_SLOT_START
```
I wonder if this one shouldn't be moved out of this file to the generic Kconfig.dma.

hakehuang replied:
this is only required for kinetis, which test channel starts from 58, other platfrom like RT or ST SOC starts from 0

erwango replied:
@FRASTM Didn't we discuss something similar ?

erwango replied:
> this is only required for kinetis, which test channel starts from 58, other platfrom like ST SOC starts from 0, which does not need such special configure.

In such case, it would have to be DMA_MCUX_TEST_SLOT_START to make it clear.

FRASTM replied:
In some stm32 socs, it counts from 1, some from 0.  
In the test application chan_blen_transfer, I had a loop on the chan_id to select one or a more channel to test on. Something like :
```
#define FIRST_DMA_CHANNEL	CONFIG_FIRST_DMA_CHANNEL
#define LAST_DMA_CHANNEL	CONFIG_LAST_DMA_CHANNEL

	for (chan_id = FIRST_DMA_CHANNEL;
	     chan_id <= LAST_DMA_CHANNEL;
	     chan_id++) {
		if (dma_config(dma, chan_id, &dma_cfg)) {
			printk("ERROR: transfer config\n");
			return;
		}
}
```

hakehuang replied:
I will change to DMA_MCUX_TEST_SLOT_START

erwango commented on include/drivers/dma.h:
```diff
@@ -31,7 +31,8 @@ extern "C" {
 	MEMORY_TO_PERIPHERAL,
-	PERIPHERAL_TO_MEMORY
+	PERIPHERAL_TO_MEMORY,
+	PERIPHERAL_TO_PERIPHERAL /*only supported in NXP EDMA*/
```
If specific to NXP, some checks should be added in other vendors drivers to be sure it will not be used

erwango replied:
After quick check only one will not report errors IIC: `drivers/dma/dma_sam0.c`

hakehuang replied:
so need I update the dma_sam0.c driver for this? @erwango

erwango replied:
I'd say yes, please.

hakehuang replied:
updated

erwango commented on tests/drivers/dma/chan_blen_transfer/src/test_dma.c:
```diff
@@ -93,7 +108,7 @@ void test_dma_m2m_chan0_burst8(void)
 void test_dma_m2m_chan1_burst8(void)
 {
-	zassert_true((test_task(1, 8) == TC_PASS), NULL);
+	zassert_true((test_task(0, 8) == TC_PASS), NULL);
```
I don't this this is a clean change.
IIUC, this makes `test_dma_m2m_chan0_burst8` testing the same as `test_dma_m2m_chan1_burst8`.
So chan0 is not tested anymore fro those who need it

FRASTM replied:
It could be a test_task(FIRST_DMA_CHANNEL, 8) and another one test_task(LAST_DMA_CHANNEL, 8) (idem for burst16)

hakehuang replied:
sorry I forget to revert my debug change, change it back

hakehuang committed:
dma: add error check on peripheral to peripheral

add dma direction check for dma_sam0

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

erwango reviewed (approved):
I'm fine with current change, but it would have been nice that others can use DMA_MCUX_TEST_SLOT_START.  This can be changed in upcoming change.

erwango commented on drivers/dma/dma_sam0.c:
```diff
@@ -149,10 +149,14 @@ static int dma_sam0_config(struct device *dev, u32_t channel,
 		chcfg->CHCTRLA.reg = DMAC_CHCTRLA_TRIGACT_TRANSACTION |
 				     DMAC_CHCTRLA_TRIGSRC(config->dma_slot);
-	} else {
+	} else if (config->channel_direction == MEMORY_TO_PERIPHERAL ||
```
Please put operands inside brackets

hakehuang replied:
ok

carlescufi commented:
API meeting:

No objections to the changes in the DMA config structure, except for additional comments.

carlescufi commented on include/drivers/dma.h:
```diff
@@ -98,25 +99,28 @@ struct dma_block_config {
  * @brief DMA configuration structure.
  *
- *     dma_slot             [ 0 : 5 ]   - which peripheral and direction
+ *     dma_slot             [ 0 : 6 ]   - which peripheral and direction
```
Can you please explain why this grew by one bit?

hakehuang replied:
ok, in Imx.rt series the dma slot are 127 so I need  to expand them. I will add it in the commit message

carlescufi commented on include/drivers/dma.h:
```diff
@@ -98,25 +99,28 @@ struct dma_block_config {
  *                                        chaining.
  *                                        0-disable, 1-enable
- *     reserved             [ 19 : 31 ]
+ *     linked_channel       [ 20 : 26 ] - linked channel number
```
What is this exactly, how is it used and does it apply to all DMA drivers?

hakehuang replied:
in imx.rt we can interlink channel, and as there are 127 channels so we need 7 bits for this link_channel settings.  Ohter drivers will not use it as this is used to be a reserved bits.

MaureenHelm commented:
`tests/drivers/dma/loop_transfer` doesn't seem to work at runtime:

On `frdm_k64f`:
```
*** Booting Zephyr OS build v2.3.0-rc1-6-g90ad28307ca0  ***
DMA memory to memory transfer started on DMA_0
Preparing DMA Controller
Starting the transfer and waiting for 1 second
TX data: The quick brown fox jumps over the lazy dog
block_size 43
I: channel 0 error status is 0x2
ERROR: unfinished transfer
Each RX buffer should contain the full TX buffer string.
TX data: The quick brown fox jumps over the lazy dog
RX data Loop 0: The quick brown fox jumps over the lazy dog
RX data Loop 1: The quick brown fox jups
RX data Loop 2: 
RX data Loop 3: 
RX data Loop 4: 
Finished: DMA
```

On `mimxrt1064_evk`:
```
*** Booting Zephyr OS build v2.3.0-rc1-6-g90ad28307ca0  ***
DMA memory to memory transfer started on DMA_0
Preparing DMA Controller
Starting the transfer and waiting for 1 second
TX data: The quick brown fox jumps over the lazy dog
block_size 43
Each RX buffer should contain the full TX buffer string.
TX data: The quick brown fox jumps over the lazy dog
RX data Loop 0: The quick brown fox jumps over the lazy dogpqt
RX data Loop 1: The quick brown fox jumps over the lazy dog
RX data Loop 2: The quick brown fox jumps over the lazy dogp
RX data Loop 3: The quick brown fox jumps over the lazy dog
RX data Loop 4: The quick brown fox jumps over the lazy dog�
Finished: DMA
```

hakehuang commented:
@MaureenHelm ,  I see problem in FRDM_K64F, I will debug this. for RT1064 problem, it is because  the buffer is more than the data transferred, and some ramdon data is print after dog, let me fix it as well

hakehuang committed:
dt-bindings: dts/arm/nxp : add MCUX EDMA Support

add MCUX EDMA support for dts binding

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
dma: add EDMA MCUX support for RT and k6s

tested on mimxrt1060_evt
  MEMORY_NOCACHE is needed
test on frdmk64f
  special test slot need configure with
  CONFIG_DMA_TEST_SLOT_START

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
